### PR TITLE
Add KSWIN as a concept drift detector

### PR DESCRIPTION
**Pull request by ChristophRaab**
Changes proposed in this pull request:

*  Added new KSWIN concept drift detector based on [Raab et al. (2020)](https://www.sciencedirect.com/science/article/pii/S0925231220305063) with documentation
*  Added tests for KSWIN concept drift detector

> Christoph Raab, Moritz Heusinger, Frank-Michael Schleif,
> Reactive Soft Prototype Computing for Concept Drift Streams,
> Neurocomputing, 2020

---

#### Implementation
- [X] Implementation is correct (it performs its intended function).
- [X] Code complies with PEP-8 and is consistent with the framework.
- [X] Code is properly documented.
- [X] PR description covers ALL the changes performed.
- [X] Files changed (update, add, delete) are in the PR's scope (no extra files are included).

#### Tests
- [X] New functionality is tested.
- [X] Tests are created for the new functionality or existing tests are updated accordingly.
- [X] ALL tests pass with no errors.
- [X] CI/CD pipelines run with no errors.
- [X] Test Coverage is maintained (coverage may drop by no more than 0.2%).

### CODE REVIEW

**Comment by jacobmontiel** [comment]
Thanks for submitting this PR @ChristophRaab 

Is this implementation the same as the one used for the paper?

**Comment by ChristophRaab** [comment]
Yes, the KSWIN detector is proposed within the publication. Feel free to look it up here: https://doi.org/10.1016/j.neucom.2019.11.111

**Comment by jacobmontiel** [comment]
Sorry, my question was not very clear. I meant to ask if the results reported in the paper were obtained with this implementation. This is to asses the correctness of the implementation apart from the functional tests already included in the PR.

**Comment by ChristophRaab** [comment]
No problem. Yes, with this implementation the results in the paper are obtained. 
It is the same as in the repository of the publication:
https://github.com/ChristophRaab/rrslvq/blob/master/kswin.py

**Comment by ChristophRaab** [comment]
Sorry if i don't get it, but is there anything left for me to to?

**Comment by jacobmontiel** [comment]
> Sorry if i don't get it, but is there anything left for me to do?

@ChristophRaab sorry for the delay, I will start the review and get back with some comments.

File: tests/drift_detection/test_kswin.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    pre obtained data initialisation test.
+    data must be list
+
+    KSWIN window size initialisation test.
+    0 < stat_size <  w_size
+
+    KSWIN change detector size initialisation test.
+    At least 1 false positive must arisie due to the sensitive alpha, when testing the standard
+    Sea generator
+    """
```
Comment (jacobmontiel) [inline tests/drift_detection/test_kswin.py:21]: Please use the `pytest` mechanism for handling `exceptions` in tests:

https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions

File: tests/drift_detection/test_kswin.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        assert False
+    try:
+        KSWIN(w_size=10, stat_size=30)
+    except ValueError:
+        assert True
+    else:
+        assert False
+
+    kswin = KSWIN(alpha=0.001)
+    stream = SEAGenerator(classification_function=2,\
```
Comment (jacobmontiel) [inline tests/drift_detection/test_kswin.py:56]: Perhaps the [`ConceptDriftStream`](https://scikit-multiflow.github.io/scikit-multiflow/api/generated/skmultiflow.data.ConceptDriftStream.html#skmultiflow.data.ConceptDriftStream) generator is better suited for this case, to simulate drift in the stream.

File: tests/drift_detection/test_kswin.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    for i in range(1000):
+        data = stream.next_sample(10)
+        batch = data[0][0][0]
+        mean.append(batch)
+        kswin.add_element(batch)
+        if kswin.detected_change():
+            mean = []
+            detections.append(i)
+    assert len(detections) > 1
```
Comment (jacobmontiel) [inline tests/drift_detection/test_kswin.py:69]: This test is too weak as it does not protect against changes in the behavior of the method. In other drift detectors, we keep track of the sample index where drift is detected.

File: src/skmultiflow/drift_detection/kswin.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+    data: list (default=None,optional)
+        Already collected data to avoid cold start.
+
+    Notes
+    -----
+    KSWIN (Kolmogorov-Smirnov Windowing) [1]_ is a concept change detection method based
+    on the Kolmogorov-Smirnov (KS) statistical test. KS-test is a statistical test with
+    no assumption of underlying data distribution. KSWIN can monitor data or performance
+    distributions. Note that testing is always dimensions-wise.
```
[the remark this answers is not in the document]
Comment (ChristophRaab) [reply src/skmultiflow/drift_detection/kswin.py:29]: Sorry for the confusion. The detector simply accepts one dimensinal data. i will change it in the upcoming commit.

File: src/skmultiflow/drift_detection/kswin.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    def reset(self):
+        """ reset
+
+        Resets the change detector parameters.
+        """
+        self.alpha = 0
+        self.window = []
+        self.change_detected = False
+
+    def get_info(self):
```
[the remark this answers is not in the document]
Comment (ChristophRaab) [reply src/skmultiflow/drift_detection/kswin.py:147]: As it is redundant, the method is removed.

File: src/skmultiflow/drift_detection/kswin.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+            raise ValueError("Alpha must be between 0 and 1")
+
+        if self.w_size < 0:
+            raise ValueError("w_size must be greater than 0")
+
+        if self.w_size < self.stat_size:
+            raise ValueError("stat_size must be smaller than w_size")
+
+        if type(data) != list or type(data) is None:
+            self.window = []
```
[the remark this answers is not in the document]
Comment (ChristophRaab) [reply src/skmultiflow/drift_detection/kswin.py:93]: As you suggested above, the window list is now numpy array. And the queues functionality is easily implemented.

File: tests/drift_detection/test_kswin.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        assert True
+    else:
+        assert False
+    try:
+        KSWIN(w_size=10, stat_size=30)
+    except ValueError:
+        assert True
+    else:
+        assert False
+
```
[the remark this answers is not in the document]
Comment (ChristophRaab) [reply tests/drift_detection/test_kswin.py:54]: I split the test between initialization and functionality as you suggested

File: tests/drift_detection/test_kswin.py (hunk 3 above)
Comment (ChristophRaab) [reply tests/drift_detection/test_kswin.py:69]: I adopted the test of Adwin and other detectors to check the behaviour of KSWIN.

File: tests/drift_detection/test_kswin.py (hunk 2 above)
Comment (ChristophRaab) [reply tests/drift_detection/test_kswin.py:56]: This is true, but i adopted the test of Adwin and other detectors to be consistent with the other tests.

**Comment by ChristophRaab** [comment]
See [27cf7f8](https://github.com/scikit-multiflow/scikit-multiflow/pull/234/commits/27cf7f8c144f7a35bc52ca94d46e52466cfa8f3d) for the change to the pull request. 

The following table shows the performance of drift detectors on the concept drift streams. 
performance_kswin

**Pull request merged**